avro/src/main/java/org/apache/apex/malhar/avro/AvroFileInputOperator.java

```diff
@@ -0,0 +1,172 @@
+import org.apache.avro.io.DatumReader;
+import org.apache.hadoop.classification.InterfaceStability;
+import org.apache.hadoop.fs.Path;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import com.datatorrent.api.AutoMetric;
+import com.datatorrent.api.DefaultOutputPort;
+import com.datatorrent.lib.io.fs.AbstractFileInputOperator;
+
+/**
+ * <p>
+ * Avro File Input Operator
+ * </p>
+ * A specific implementation of the AbstractFileInputOperator to read Avro
+ * container files.<br>
+ * No need to provide schema,its inferred from the file<br>
+ * This operator emits a GenericRecord based on the schema derived from the
+ * input file<br>
+ * Users can add the {@link FSWindowDataManager}
+ * to ensure exactly once semantics with a HDFS backed WAL.
+ *
+ * @displayName AvroFileInputOperator
+ * @category Input
+ * @tags fs, file,avro, input operator
+ *
+ * @since
+ */
+@InterfaceStability.Evolving
+public class AvroFileInputOperator extends AbstractFileInputOperator<GenericRecord>
+{
+
+  private transient long offset = 0L;
+
+  @AutoMetric
+  @VisibleForTesting
+  int recordCount = 0;
+
+  @AutoMetric
+  @VisibleForTesting
+  int errorCount = 0;
+
+  private transient DataFileStream<GenericRecord> avroDataStream;
+
+  public final transient DefaultOutputPort<GenericRecord> output = new DefaultOutputPort<GenericRecord>();
+
+  public final transient DefaultOutputPort<String> completedFilesPort = new DefaultOutputPort<String>();
+
+  public final transient DefaultOutputPort<String> errorRecordsPort = new DefaultOutputPort<String>();
+
+  /**
+   * Returns a input stream given a file path
+   *
+   * @param path
+   * @return InputStream
+   * @throws IOException
+   */
+  @Override
+  protected InputStream openFile(Path path) throws IOException
+  {
+    InputStream is = super.openFile(path);
+    if (is != null) {
+      DatumReader<GenericRecord> datumReader = new GenericDatumReader<GenericRecord>();
+      avroDataStream = new DataFileStream<GenericRecord>(is, datumReader);
+      datumReader.setSchema(avroDataStream.getSchema());
+    }
+    return is;
+  }
+
+  /**
+   * Reads a GenericRecord from the given input stream<br>
+   * Emits the FileName,Offset,Exception on the error port if its connected
+   *
+   * @return GenericRecord
+   */
+  @Override
+  protected GenericRecord readEntity() throws IOException
+  {
+    GenericRecord record = null;
+
+    record = null;
+
+    try {
+      if (avroDataStream != null && avroDataStream.hasNext()) {
+        offset++;
+        record = avroDataStream.next();
+        recordCount++;
+        return record;
+      }
+    } catch (AvroRuntimeException are) {
+      LOG.error("Exception in parsing record for file - " + super.currentFile + " at offset - " + offset, are);
+      if (errorRecordsPort.isConnected()) {
+        errorRecordsPort.emit("FileName:" + super.currentFile + ", Offset:" + offset);
+      }
+      errorCount++;
+      throw new AvroRuntimeException(are);
```

Comment:
why is the original exception not thrown

Reply:
Few of these comments are on the existing code, which we were not planning to change as a part of this PR. We wanted to improve this operator step by step. Anyways, now it is up to @SaumyaMohan to decide whether she wants to continue with this PR.